email/blocker.go

```diff
@@ -0,0 +1,223 @@
+package email
+
+import (
+	"abuse-scanner/database"
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	"gitlab.com/NebulousLabs/errors"
+	"go.mongodb.org/mongo-driver/bson"
+)
+
+const (
+	// blockFrequency defines the frequency with which we scan for emails for
+	// which the parsed emails have not been blocked yet.
+	blockFrequency = 30 * time.Second
+)
+
+type (
+	// Blocker is an object that will periodically scan the database for abuse
+	// reports that have not been blocked yet.
+	Blocker struct {
+		staticBlockerApiUrl     string
+		staticBlockerAuthHeader string
+		staticContext           context.Context
+		staticDatabase          *database.AbuseScannerDB
+		staticLogger            *logrus.Entry
+		staticWaitGroup         *sync.WaitGroup
+	}
+
+	// BlockPOST is the datastructure expected by the blocker API
+	BlockPOST struct {
+		Skylink  string                 `json:"skylink"`
+		Reporter database.AbuseReporter `json:"reporter"`
+		Tags     []string               `json:"tags"`
+	}
+)
+
+// NewBlocker creates a new blocker.
+func NewBlocker(ctx context.Context, blockerAuthHeader, blockerApiUrl string, database *database.AbuseScannerDB, waitGroup *sync.WaitGroup, logger *logrus.Logger) *Blocker {
+	return &Blocker{
+		staticBlockerAuthHeader: blockerAuthHeader,
+		staticBlockerApiUrl:     blockerApiUrl,
+		staticContext:           ctx,
+		staticDatabase:          database,
+		staticLogger:            logger.WithField("module", "Blocker"),
+		staticWaitGroup:         waitGroup,
+	}
+}
+
+// Start initializes the blocker process.
+func (b *Blocker) Start() error {
+	go b.threadedBlockMessages()
+	return nil
+}
+
+// threadedBlockMessages will periodically fetch email messages that have not
+// been blocked yet and feed them to the blocker API.
+func (b *Blocker) threadedBlockMessages() {
+	// convenience variables
+	abuseDB := b.staticDatabase
+	logger := b.staticLogger
+	wg := b.staticWaitGroup
+
+	ticker := time.NewTicker(blockFrequency)
+	first := true
+
+	// start the loop
+	for {
+		// sleep until next iteration, sleeping at the start of the for loop
+		// allows to 'continue' on error.
+		if !first {
+			select {
+			case <-b.staticContext.Done():
+				logger.Debugln("Blocker context done")
+				return
+			case <-ticker.C:
+			}
+		}
+		first = false
+
+		logger.Debugln("Triggered")
+
+		func() {
+			wg.Add(1)
+			defer wg.Done()
+
+			// fetch all unblocked emails
+			toBlock, err := abuseDB.FindUnblocked()
+			if err != nil {
+				logger.Errorf("Failed fetching unparsed emails, error %v", err)
+				return
+			}
+
+			// log unblocked messages count
+			numUnblocked := len(toBlock)
+			if numUnblocked == 0 {
+				logger.Debugf("Found %v unblocked messages\n", numUnblocked)
```

Comment:
f/u Might want to consider a `continue` here even though it's strictly speaking not necessary due to the for loop being the last thing executing before another iteration. Still nice to safe the reader scrolling down to double check. `vet` might then complain about the `else` though and suggest dropping it similar to when you add a `return`.